revocation/ocsp/ocsp.go

```diff
@@ -0,0 +1,235 @@
+func CheckStatus(opts Options) [][]error {
+	result := make([][]error, len(opts.CertChain))
+
+	if len(opts.CertChain) == 0 {
+		return result
+	}
+
+	// Validate cert chain structure
+	if len(opts.CertChain) == 1 {
+		if err := validateIssuer(opts.CertChain[0], opts.CertChain[0]); err != nil {
+			fillResultsWithError(result, InvalidChainError{IsInvalidRoot: true, Err: err})
+			return result
+		}
+	} else {
+		if err := coreX509.ValidateCodeSigningCertChain(opts.CertChain, &opts.SigningTime); err != nil {
+			fillResultsWithError(result, InvalidChainError{Err: err})
+			return result
+		}
+	}
+
+	// Check status for each cert in cert chain
+	var wg sync.WaitGroup
+	for i, cert := range opts.CertChain[:len(opts.CertChain)-1] {
+		wg.Add(1)
+		// Assume cert chain is accurate and next cert in chain is the issuer
+		go func(i int, cert *x509.Certificate) {
+			defer wg.Done()
+			result[i] = certCheckStatus(cert, opts.CertChain[i+1], opts)
+		}(i, cert)
+	}
+	// Last is root cert, which will never be revoked by OCSP
+	result[len(opts.CertChain)-1] = []error{nil}
+
+	wg.Wait()
+	return result
+}
+
+func fillResultsWithError(results [][]error, err error) {
+	for i := range results {
+		results[i] = []error{err}
+	}
+}
+
+func validateIssuer(subject *x509.Certificate, issuer *x509.Certificate) error {
+	if err := subject.CheckSignatureFrom(issuer); err != nil {
+		return err
+	}
+	if !bytes.Equal(issuer.RawSubject, subject.RawIssuer) {
+		return errors.New("parent's subject does not match issuer for a cert in the chain")
+	}
+	return nil
+}
+
+func certCheckStatus(cert, issuer *x509.Certificate, opts Options) []error {
+	ocspURLs := cert.OCSPServer
+	if len(ocspURLs) == 0 {
+		// OCSP not enabled for this certificate.
+		return []error{NoOCSPServerError{}}
+	}
+
+	serverErrs := make([]error, len(ocspURLs))
+	for serverIndex, server := range ocspURLs {
+		err := checkStatusFromServer(cert, issuer, server, opts)
+		switch t := err.(type) {
+		case nil, RevokedError, UnknownStatusError:
+			// A valid response has been received from an OCSP server
+			// Result should be based on only this response, not any errors from
+			// other servers
+			return []error{t}
+		default:
+			serverErrs[serverIndex] = err
+			continue
+		}
+
+	}
+	return serverErrs
+}
+
+func checkStatusFromServer(cert, issuer *x509.Certificate, server string, opts Options) error {
+	// Check valid server
+	if serverURL, err := url.Parse(server); err != nil || strings.ToLower(serverURL.Scheme) != "http" {
+		// This function is only able to check servers that are accessible via HTTP
+		return OCSPCheckError{Err: errors.New("OCSPServer must be accessible over HTTP")}
+	}
+
+	// Create OCSP Request
+	resp, err := executeOCSPCheck(cert, issuer, server, opts)
+	if err != nil {
+		// If there is a server error, attempt all servers before determining what to return
+		// to the user
+		if errors.Is(err, TimeoutError{}) {
+			return err
+		}
+		return OCSPCheckError{Err: err}
```

Comment:
Can we move this logic to executeOCSPCheck?

Reply:
This was just a quick way to avoid wrapping all the `return err` in executeOCSPCheck in an OCSPCheckError to increase readability, but I don't have any objection to wrapping them and then just doing return err here